pkg/controllers/management/clusterupstreamrefresher/cluster_upstream_refresher.go

```diff
@@ -0,0 +1,269 @@
+package clusterupstreamrefresher
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"strconv"
+	"strings"
+	"time"
+
+	eksv1 "github.com/rancher/eks-operator/pkg/apis/eks.cattle.io/v1"
+	apimgmtv3 "github.com/rancher/rancher/pkg/apis/management.cattle.io/v3"
+	v3 "github.com/rancher/rancher/pkg/generated/controllers/management.cattle.io/v3"
+	mgmtv3 "github.com/rancher/rancher/pkg/generated/norman/management.cattle.io/v3"
+	"github.com/rancher/rancher/pkg/settings"
+	"github.com/rancher/rancher/pkg/wrangler"
+	wranglerv1 "github.com/rancher/wrangler/pkg/generated/controllers/core/v1"
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+const (
+	noKEv2Provider         = "none"
+	clusterLastRefreshTime = "clusters.management.cattle.io/ke-last-refresh"
+)
+
+type clusterRefreshController struct {
+	secretsCache        wranglerv1.SecretCache
+	clusterClient       v3.ClusterClient
+	clusterCache        v3.ClusterCache
+	clusterEnqueueAfter func(name string, duration time.Duration)
+}
+
+// for other cloud drivers, please edit HERE
+type clusterConfig struct {
+	eksConfig *eksv1.EKSClusterConfigSpec
+}
+
+func Register(ctx context.Context, wContext *wrangler.Context) {
+	c := clusterRefreshController{
+		secretsCache:        wContext.Core.Secret().Cache(),
+		clusterClient:       wContext.Mgmt.Cluster(),
+		clusterCache:        wContext.Mgmt.Cluster().Cache(),
+		clusterEnqueueAfter: wContext.Mgmt.Cluster().EnqueueAfter,
+	}
+
+	wContext.Mgmt.Cluster().OnChange(ctx, "cluster-refresher-controller", c.onClusterChange)
+}
+
+func (c *clusterRefreshController) onClusterChange(key string, cluster *mgmtv3.Cluster) (*mgmtv3.Cluster, error) {
+	if cluster == nil || cluster.DeletionTimestamp != nil {
+		return cluster, nil
+	}
+
+	provider, ready := getProviderAndReadyStatus(cluster)
+	if provider == noKEv2Provider {
+		// not a KEv2 cluster
+		return cluster, nil
+	}
+
+	if !ready {
+		return cluster, nil
+	}
+
+	var lastRefreshTime string
+	var err error
+
+	if cluster.Annotations != nil {
+		lastRefreshTime = cluster.Annotations[clusterLastRefreshTime]
+	}
+
+	providerRefreshInterval, err := getProviderRefreshInterval(provider)
+	if err != nil {
+		return cluster, err
+	}
+
+	nextRefreshTime, err := getNextProviderRefreshTime(providerRefreshInterval, lastRefreshTime)
+	if err != nil {
+		return cluster, err
+	}
+
+	if !time.Now().After(nextRefreshTime) {
+		return cluster, err
+	}
```

Comment:
The last time.Now check should move inside getNextProviderRefreshTime then rename getNextProviderRefreshTime to something like `shouldRefresh` and return a bool and error. This logic fits better inside shouldRefresh and nextRefreshTime isn't used here.

Reply:
I moved it to `func shouldRefreshCluster()`